### PR TITLE
Just some tbl things.

### PR DESCRIPTION
**Pull request by Drup**
I always felt that the `Tbl` module was a ~~gross~~slight violation of the DRY principle, so this patchset removes it.
On the way, I removed almost all the duplicated `(String|Int)(Map|Set)` in the compiler, and a few others.

This is my small contribution towards making the compiler feel like one code base, instead of the collision of five different ones. ;)

## Details on `Tbl`

`Tbl` is basically a polymorphic `Map` module, using the polymorphic comparison with a few more hacks (such as `Tbl.find_str` which hardcodes the string comparison for speed. It is used for some string tables ... but not all of them). It has the exact same structure and semantics as Map, making it completely redundant.

On top of that, relying on the generic Tbl make local modifications fairly painful and invasive. For example, in the eliom patches, names are a couple of a string and a "location" (in eliom vocabulary) with a custom comparison function. With this patch, it can be implemented as a simple change in the implementation of `Env.NameMap`.

## Patchset

The first 10 commits should be fairly consensual. They were mostly done through paste and replace.

The last 3 commits, particularly 484c79b , might be less consensual: Tbl made its way into the format for cmo files though the symbol tables for global variables and primitives. I replaced them by an `Ident.Map` (for global ids) and a `StringMap` (for primitives). I didn't bump the magic number (yet).

I'm not sure if I need a change entry for this. 

## Perfs

I don't expect any perf differences (the "hot" paths were already using `Tbl.find_str`). My non-scientific benchmark show a very minor speedup (around 1%, so probably noise). Is there some command to measure performance changes of the compiler on itself?

### CODE REVIEW

**Comment by Drup** [comment]
I do not understand travis' failure. It builds/tests fine locally and the various `depend` are up-to-date.

**Comment by Octachron** [comment]
There is a `.depend` file hidden in ocamldoc/stdlib_non_prefixed, it seems that this the one responsible for the failure.

**Comment by mshinwell** [comment]
I haven't got time to read this in detail right now, but will aim to do so in due course.  I agree that cleanup here would be beneficial.  This work seems related to some changes I am working on to integrate the `utils/identifiable.ml*` functionality directly into the standard library.  I will write about that within the next few weeks.

**Comment by nojb** [comment]
>  I didn't bump the magic number (yet).

This is not necessary. Magic numbers will be bumped unconditionally for each release.

**Comment by dra27** [comment]
At a glance, this looks good to me (although, if I may, I think the word *gross* should possibly only be used when describing one's own work!).

It's a basically internal change, so I agree it doesn't *require* a changes entry, but that does mean the work only gets credited in Git history, so feel free to add one if you'd like!

**Comment by Drup** [comment]
> This work seems related to some changes I am working on to integrate the utils/identifiable.ml* functionality directly into the standard library.

Indeed! Making relevant data-structures in the compiler into `Identifiable.S` would also be very nice. I almost started doing that when I realized that `Identifiable` also include printing, and printing for the source/type language is not so trivial (the printers are defined somewhere else, and require lot's of machinery).

> It's a basically internal change, so I agree it doesn't require a changes entry, but that does mean the work only gets credited in Git history, so feel free to add one if you'd like!

Ok, I will add one then.

File: asmcomp/closure.ml [hunk 1]
```diff
@@ -16,12 +16,14 @@
 (* Introduction of closures, uncurrying, recognition of direct calls *)
 
 open Misc
+open Numbers
```
Comment (mshinwell) [inline asmcomp/closure.ml]: Instead of a whole `open`, I would just use: `module Int = Numbers.Int` here.

File: asmcomp/closure.ml [hunk 2]
```diff
@@ -16,12 +16,14 @@
...
 open Asttypes
 open Primitive
 open Lambda
 open Switch
 open Clambda
 
+module VarMap = Ident.Map
```
Comment (mshinwell) [inline asmcomp/closure.ml]: I think this is confusing ("ident"s versus "var"s).  How about just replacing VarMap with Ident.Map throughout?  That seems fine to me, and more clear in fact.

File: asmcomp/cmmgen.ml [hunk 3]
```diff
@@ -3397,11 +3391,7 @@ let curry_function arity =
   else [tuplify_function (-arity)]
 
 
-module IntSet = Set.Make(
-  struct
-    type t = int
-    let compare (x:t) y = compare x y
-  end)
+module IntSet = Numbers.Int.Set
```
Comment (mshinwell) [inline asmcomp/cmmgen.ml]: Here, I would use `module Int = Numbers.Int` again, then this alias can be removed.  Just use `Int.Set` instead of `IntSet`.

File: asmcomp/selectgen.ml [hunk 4]
```diff
@@ -17,32 +17,33 @@
    sequentialization. *)
 
 open Misc
+open Numbers
```
Comment (mshinwell) [inline asmcomp/selectgen.ml]: Same comment as above.

File: bytecomp/simplif.ml [hunk 5]
```diff
@@ -19,6 +19,8 @@
 open Asttypes
 open Lambda
 
+module VarMap = Ident.Map
```
Comment (mshinwell) [inline bytecomp/simplif.ml]: Same comment as above.

File: driver/makedepend.ml [hunk 6]
```diff
@@ -13,9 +13,9 @@
 (*                                                                        *)
 (**************************************************************************)
 
+open Misc
```
Comment (mshinwell) [inline driver/makedepend.ml]: I suggest in `Misc` you add `Stdlib.String` and then (either via `open` or otherwise) use `String.Map` and `String.Set` instead of `StringMap` and `StringSet`; those latter two aliases in `Misc` could then be removed.  I would hope this change converges on a world where the stdlib modules gain support for sets and maps, and uniform naming can be used throughout.  (I have a patch for such stdlib support under development as I mentioned.)

File: bytecomp/symtable.ml [hunk 7]
```diff
@@ -30,73 +32,84 @@ type error =
 
 exception Error of error
 
-let empty_numtable = { num_cnt = 0; num_tbl = Tbl.empty }
+module NumTbl (M : Map.S) = struct
```
Comment (mshinwell) [inline bytecomp/symtable.ml]: General point: I think the usual style is to have a single capital letter for module names and then lowercase afterwards separated by underscores.  (This also means that if you move a module out into a file of the appropriate name, then you can keep using lowercase filenames as is conventional, yet not have to rename any of the module references in the code.)

File: ocamltest/ocamltest_stdlib.ml [hunk 8]
```diff
@@ -177,8 +177,8 @@ module Sys = struct
 end
 
 module StringSet = struct
-  include Set.Make (String)
+  include Misc.StringSet
   let string_of_stringset s = String.concat ", " (elements s)
```
Comment (mshinwell) [inline ocamltest/ocamltest_stdlib.ml]: (Following comment above.)  This should probably be [`Misc.Stdlib.`]`String.Set.to_string`.

File: ocamltest/ocamltest_stdlib.mli [hunk 9]
```diff
@@ -62,4 +62,4 @@ module StringSet : sig
   val string_of_stringset : t -> string
 end
 
-module StringMap : Map.S with type key = string
+module StringMap = Misc.StringMap
```
Comment (mshinwell) [inline ocamltest/ocamltest_stdlib.mli]: Why does this alias need to be re-exported?

Comment (mshinwell) [reply ocamltest/ocamltest_stdlib.mli]: (Maybe `Ocamltest_stdlib` could just be `Misc.Stdlib`?)

File: otherlibs/unix/.depend [hunk 10]
```diff
@@ -77,10 +77,10 @@ cstringv.o: cstringv.c ../../byterun/caml/mlvalues.h \
  ../../byterun/caml/freelist.h ../../byterun/caml/minor_gc.h \
  ../../byterun/caml/address_class.h ../../byterun/caml/osdeps.h \
  ../../byterun/caml/memory.h unixsupport.h
-dup.o: dup.c ../../byterun/caml/mlvalues.h ../../byterun/caml/config.h \
+dup2.o: dup2.c ../../byterun/caml/mlvalues.h ../../byterun/caml/config.h \
  ../../byterun/caml/m.h ../../byterun/caml/s.h ../../byterun/caml/misc.h \
```
Comment (mshinwell) [inline otherlibs/unix/.depend]: This appears to be an unrelated change.

File: tools/Makefile [hunk 11]
```diff
@@ -287,11 +287,13 @@ install::
 
 # The bytecode disassembler
 
-DUMPOBJ=opnames.cmo dumpobj.cmo
+DUMPOBJ= \
+          ../compilerlibs/ocamlcommon.cma \
+          ../compilerlibs/ocamlbytecomp.cma \
+          \
```
Comment (mshinwell) [inline tools/Makefile:308]: Are such blank lines really necessary?

Comment (mshinwell) [reply tools/Makefile:308]: (In fact maybe you could pull this change---the one for `tools/Makefile` that is---into a separate GPR?)

File: typing/env.ml [hunk 12]
```diff
@@ -150,7 +150,7 @@ end  = struct
 
 end
 
-module PathMap = Map.Make(Path)
+module NameMap = Misc.StringMap
```
Comment (mshinwell) [inline typing/env.ml]: Names of what?

File: typing/includemod.ml [hunk 13]
```diff
@@ -165,6 +164,12 @@ let kind_of_field_desc = function
   | Field_class _ -> "class"
   | Field_classtype _ -> "class type"
 
+(** Map indexed field names. Allow to preserve namespaces. *)
```
Comment (mshinwell) [inline typing/includemod.ml]: This comment isn't very clear---can you reword?

File: typing/includemod.ml [hunk 14]
```diff
@@ -509,9 +514,9 @@ and check_modtype_equiv ~loc env ~mark cxt mty1 mty2 =
 
 let can_alias env path =
   let rec no_apply = function
-    | Pident _ -> true
-    | Pdot(p, _, _) -> no_apply p
-    | Papply _ -> false
+    | Path.Pident _ -> true
+    | Path.Pdot(p, _, _) -> no_apply p
+    | Path.Papply _ -> false
```
Comment (mshinwell) [inline typing/includemod.ml:522]: FWIW, using disambiguation helps a lot for this kind of thing:
```
let can_alias env path =
  let rec no_apply (path : Path.t) =
    match path with
 ￼   | Pident _ -> true
    | Pdot(p, _, _) -> no_apply p
    | Papply _ -> false
  in
  ...
```

File: typing/path.ml [hunk 15]
```diff
@@ -107,3 +107,10 @@ let is_constructor_typath p =
   match constructor_typath p with
   | Regular _ -> false
   | _ -> true
+
+module T = struct
+  type nonrec t = t
+  let compare = compare
+end
+module Set = Set.Make(T)
+module Map = Map.Make(T)
```
Comment (mshinwell) [inline typing/path.ml:116]: You could probably use `Identifiable.Make` for this instead.

File: typing/typetexp.ml [hunk 16]
```diff
@@ -68,8 +68,10 @@ type error =
 exception Error of Location.t * Env.t * error
 exception Error_forward of Location.error
 
+(** Map indexed by type variable names. *)
+module VarMap = Misc.StringMap
```
Comment (mshinwell) [inline typing/typetexp.ml]: Here is an example of why I recommended not having the `VarMap` alias up above in `closure.ml`: here the same alias means something completely different!  (And it's not like it's a one-letter alias.)

**Review by mshinwell — CHANGES_REQUESTED** [review]
I've made some comments and suggestions.  This is good work.

**Comment by mshinwell** [comment]
Oh, and please add a `Changes` entry, in the "internal" section.  This is a fairly large diff even if mostly straightforward.

File: otherlibs/unix/.depend (hunk 10 above)
Comment (Drup) [reply otherlibs/unix/.depend]: I just made `make alldepend`. Do you want me to revert that bit?

File: tools/Makefile (hunk 11 above)
Comment (Drup) [reply tools/Makefile:308]: I'm not sure what you mean. With this patch, dumpobj depends on `Misc`. I could try to finetune the link dependencies, but I really don't see the point for an internal tool.

File: typing/includemod.ml (hunk 14 above)
Comment (Drup) [reply typing/includemod.ml:522]: Unfortunately, this part of the compiler has warning 40 turned on (and fatal). I think changing default warnings in the implementation of the typechecker shouldn't be discussed in this PR.

File: typing/path.ml (hunk 15 above)
Comment (Drup) [reply typing/path.ml:116]: `Identifiable.Make` require (two) printers and an equal function. I would prefer to leave that for later.

File: typing/env.ml (hunk 12 above)
Comment (Drup) [reply typing/env.ml]: `Name`, as in the individual parts in an `Longindent.t`. I would like to call them ident, but that clashes with the individual parts of a `Path.t`. I insist on having a local redefinition on that one, as it's the original motivation for this whole patch. But I agree, the name is bad.

File: typing/includemod.ml (hunk 13 above)
Comment (Drup) [reply typing/includemod.ml]: The new one is : 
```
(** Map indexed by both field types and names. 
    This avoids name clashes between different sorts of fields 
    such as values and types. *)
```

File: ocamltest/ocamltest_stdlib.mli (hunk 9 above)
Comment (Drup) [reply ocamltest/ocamltest_stdlib.mli]: Done.

File: bytecomp/symtable.ml (hunk 7 above)
Comment (Drup) [reply bytecomp/symtable.ml]: Done.

File: driver/makedepend.ml (hunk 6 above)
Comment (Drup) [reply driver/makedepend.ml]: I just did that. That was extremely boring work. O_°

File: asmcomp/cmmgen.ml (hunk 3 above)
Comment (Drup) [reply asmcomp/cmmgen.ml]: Done.

File: typing/typetexp.ml (hunk 16 above)
Comment (Drup) [reply typing/typetexp.ml]: I removed all instances of `VarMap` outside of the typechecker and renamed that one as `TyVarMap`, which should make things clearer.

**Comment by Drup** [comment]
@mshinwell Thanks for the detailed review! I think I addressed most of your comments. I also fixed the change entry and rebased.

There are a few things (cleanup the 3 extended stdlibs, add even more identifiable everywhere) that would be very beneficial, but I would prefer to do them later, as the risk of conflict between this patchset and other changes is quite high.

**Comment by Drup** [comment]
While replacing ocamltest's StringMap and Sets, I realized the function `stringset_of_string` was unused. So I removed it and simply use aliases to Map.Make's output.

**Comment by Drup** [comment]
Thanks to @kit-ty-kate, the output of the opam run is [here](http://51.145.134.72:8080/). Apparently, there is an issue with annot files.

**Comment by Drup** [comment]
I rebased and added some fixed related to various recent additions.

I think I already addressed all of @mshinwell 's comment. Is there a remaining blocker?

I plan to squash things before the merge.

**Comment by damiendoligez** [comment]
You don't have to squash, just state that it should be squashed and we just use GitHub's "Squash and merge" option.

**Comment by gasche** [comment]
"Squash and merge" is generally terrible except for PRs that obviously fit in just one commit. If it's a series of nicely separated commits ("Remove Tbl in Foo" fits the pattern) plus some junk, you have to squash manually.

On the other hand, @Drup is making a mistake by not squashing right away. He is trying to save his time, but introduces complexity, delay, communication needs and mistake points in the process instead. One might even go as far as saying as he deserves to get his entry squashed!

**Comment by Drup** [comment]
@gasche I'm going to quote @mshinwell In another PR of mine:

> Please ensure this isn't squashed (or similar) before merging as it will be difficult to review.

New commits are not squashed immediately with the old ones so that reviewers don't have to make high order diffs between stack of patchs. Unfortunately, we're not using Iron for OCaml PR reviews.

**Comment by gasche** [comment]
I would disagree with Mark on this one, but given that he is the one who should give you his seal of approval, it makes sense to please him.

**Comment by xavierleroy** [comment]
"Squash and merge" is generally the right thing to do, because most PRs have an uninteresting history and uninteresting commit message, e.g. "Typos", "Going home", or "Add a Changes entry".  Indeed this PR has one commit titled "Really?!" and another titled "Depend".  I don't want to see this nor the 18 intermediate states of this PR in OCaml's commit log. I want to see one entry in the log with one message that summarizes what the PR does in its final state.

**Comment by dra27** [comment]
If you want to tidy the history, then I'm sure @mshinwell can be persuaded to merge a clean commit history! However, before this is merged, please could check-typo violations be dealt with:

```
./asmcomp/selectgen.ml:798.81: [long-line] line is over 80 columns
./asmcomp/selectgen.ml:1129.81: [long-line] line is over 80 columns
./bytecomp/simplif.ml:419.81: [long-line] line is over 80 columns
./bytecomp/symtable.ml:43.3: [white-at-eol] whitespace at end of line
./bytecomp/symtable.ml:62.3: [white-at-eol] whitespace at end of line
./tools/dumpobj.ml:533.81: [long-line] line is over 80 columns
./typing/includemod.ml:167.48: [white-at-eol] whitespace at end of line
./typing/includemod.ml:168.64: [white-at-eol] whitespace at end of line
./typing/includemod.ml:319.81: [long-line] line is over 80 columns
./utils/misc.ml:180.40: [white-at-eol] whitespace at end of line
./utils/misc.mli:105.7: [white-at-eol] whitespace at end of line
```

**Comment by damiendoligez** [comment]
It is generally a bad idea to squash before the end because the commits get out of sync with the discussion and it becomes hard to make sense of them.

**Comment by Drup** [comment]
@dra27 All fixed, and rebased on top of trunk to take advantage of the nice typocleaning.

**Comment by dra27** [comment]
Thanks for the `check-typo` loveliness 🙂

**Comment by Drup** [comment]
4.07 is released, the world cup is finished. @mshinwell, can I get a review? :)

**Comment by mshinwell** [comment]
Please add a comment in env.ml on the "NameMap" definition to say what "Name" means here, since we can't come up with anything better.  OK to merge with that change.  Thanks for taking the time to do this, it seems like a real improvement.

**Review by mshinwell — APPROVED** [review]

**Comment by Drup** [comment]
Done, rebased, and history somewhat cleaned up.

**Comment by Drup** [comment]
CI is green.

**Comment by gasche** [comment]
I'm a bit worried by the last commit fixing a test failure. This means that the testsuite fails if I am bisecting and end in the middle of your patchset, right? Could you maybe fixup this commit at the first point where the change is required, to maintain the invariant that intermediate states are clean?

**Comment by mshinwell** [comment]
I was planning to squash the whole lot; I'm not sure any intermediate state is useful here.

**Comment by Drup** [comment]
@mshinwell Do as you wish. :p
